### PR TITLE
[Tree] [NestedSet] New additions

### PR DESCRIPTION
**Pull request by pmishev**
- [Tree] [NestedSet] Added "base" property for tree level annotation
- [Tree] [NestedSet] Added 'includeNode' option in getPathQueryBuilder() to specify whether you want the starting node included or not
- [Tree] [NestedSet] Added getPathAsString() method to entity repository
- [Tree] [NestedSet] Added 'treeRootNode' options in verify() in case you want to verify a single tree in a forest
- [Tree] [NestedSet] Added recoverFast() method for where speed is more important than safety and entity manager state
- [Tree] [NestedSet] Added options to recover() for sibling order, tree root in a forest, verification skip and auto-flushing
- [Tree] [NestedSet] Verify and recover wrong levels in nested set
- [Tree] [NestedSet] Added test for removing parent node from tree

### CODE REVIEW

File: src/Tree/Entity/Repository/NestedTreeRepository.php [hunk 1]
```diff
@@ -134,13 +134,21 @@ public function getRootNodes($sortByField = null, $direction = 'asc')
      * Get the Tree path query builder by given $node
      *
      * @param object $node
+     * @param array  $options
```
Comment (franmomu) [inline src/Tree/Entity/Repository/NestedTreeRepository.php:137]:
```suggestion
     * @phpstan-param array{includeNode?: bool}  $options
```

Can we please specify these options with array shapes? so PHPStan understands it

File: src/Tree/Entity/Repository/NestedTreeRepository.php [hunk 2]
```diff
@@ -800,10 +849,20 @@ public function reorderAll($sortByField = null, $direction = 'ASC', $verify = tr
...
      *
+     * @param array $options
+     *
+     * $options['treeRootNode']  = (object) Optional tree root node to verify, if not the whole forest (default: null)
+     *                             Option is only available for forests, not for single trees
+     *
      * @return array|bool true on success,error list on failure
      */
-    public function verify()
+    public function verify(array $options = [])
```
Comment (franmomu) [inline src/Tree/Entity/Repository/NestedTreeRepository.php:859]: I'm not using the Tree behaviour, but is it likely to pass more options? if it's just one we could maybe use `$treeRootNode`

Comment (pmishev) [reply src/Tree/Entity/Repository/NestedTreeRepository.php:859]: I agree that in this case a single `$treeRootNode` argument would be better, however I just wanted to keep consistency with other places where I added extra options. For example in the `recover()` below there are a lot optional options and one of them is again `treeRootNode`. 
In the latter case, I'd argue that it's better to have an `$options` array, at least for PHP7 (there are also other precedents in the bundle where such an `$options` array is used), but that is probably no longer the case with the named arguments in PHP8, so maybe I should use separate arguments everywhere? 
On the other hand, in the case of `getPath -> getPathQuery -> getPathQueryBuilder`, it's maybe easier to maintain with an array of options, as if you add another a new option, you don't have to change the signature of all 3 methods. WDYT?

Comment (franmomu) [reply src/Tree/Entity/Repository/NestedTreeRepository.php:859]: Hey @pmishev sorry for the late reply, yeah, I guess it's fine to keep these `$options`, please specify all these options in PHPDoc as `array{option1: type, ...}`

File: src/Tree/Entity/Repository/NestedTreeRepository.php [hunk 3]
```diff
@@ -134,13 +134,21 @@ public function getRootNodes($sortByField = null, $direction = 'asc')
...
+     * @throws InvalidArgumentException if input is not valid
+     *
      */
-    public function getPathQueryBuilder($node)
+    public function getPathQueryBuilder($node, array $options = [])
     {
+        $defaultOptions = [
+            'includeNode' => true,
+        ];
+        $options = array_merge($defaultOptions, $options);
```
Comment (phansys) [inline src/Tree/Entity/Repository/NestedTreeRepository.php:150]:
```suggestion
        $options += $defaultOptions;
```

File: src/Tree/Entity/Repository/NestedTreeRepository.php [hunk 4]
```diff
@@ -172,24 +185,60 @@ public function getPathQueryBuilder($node)
...
+     * @param object $node
+     * @param array  $options
+     *
+     * $options['includeNode']  = (bool) Whether to include the node itself (default: true)
+     * $options['separator']    = (string) The string separating the nodes of the tree
+     * $options['stringMethod'] = Entity method returning its displayable name.
+     *                            If not provided, entity must have __toString method
+     * @return string
+     */
+    public function getPathAsString($node, array $options = [])
```
Comment (phansys) [inline src/Tree/Entity/Repository/NestedTreeRepository.php:226]:
```suggestion
     */
    public function getPathAsString($node, array $options = []): string
```

File: src/Tree/Entity/Repository/NestedTreeRepository.php [hunk 5]
```diff
@@ -172,24 +185,60 @@ public function getPathQueryBuilder($node)
...
+    {
+        $defaultOptions = [
+            'includeNode'  => true,
+            'separator'    => ' > ',
+            'stringMethod' => null,
+        ];
+        $options = array_merge($defaultOptions, $options);
```
Comment (phansys) [inline src/Tree/Entity/Repository/NestedTreeRepository.php:233]:
```suggestion
        $options += $defaultOptions;
```

File: src/Tree/Entity/Repository/NestedTreeRepository.php [hunk 6]
```diff
@@ -172,24 +185,60 @@ public function getPathQueryBuilder($node)
...
+
+        $path = [];
+        $pathNodes = $this->getPath($node, $options);
+        foreach ($pathNodes as $pathNode) {
```
Comment (phansys) [inline src/Tree/Entity/Repository/NestedTreeRepository.php:237]:
```suggestion
        foreach ($this->getPath($node, $options) as $pathNode) {
```

**Review by phansys — CHANGES_REQUESTED** [review]

File: src/Tree/Entity/Repository/NestedTreeRepository.php [hunk 7]
```diff
@@ -824,25 +887,120 @@ public function verify()
...
+    {
+        $defaultOptions = [
+            'sortByField'   => null,
+            'sortDirection' => 'ASC',
+        ];
+        $options = array_merge($defaultOptions, $options);
+
+        $meta = $this->getClassMetadata();
+        $config = $this->listener->getConfiguration($this->getEntityManager(), $meta->name);
+        $self = $this;
```
[the remark this answers is not in the document]
Comment (pmishev) [reply src/Tree/Entity/Repository/NestedTreeRepository.php:925]: We don't. I've just done it the same way as in the existing `recover()` method where it's not necessary either. I'll remove it in both places.

File: src/Tree/Entity/Repository/NestedTreeRepository.php [hunk 8]
```diff
@@ -824,25 +887,120 @@ public function verify()
...
+                       ->set('node.'.$config['left'], ':left')
+                       ->set('node.'.$config['right'], ':right')
+                       ->where('node.id = :id');
+        if (isset($config['level'])) {
+            $updateQb->set('node.'.$config['level'], ':level');
+        }
+
+        $doRecover = function ($root, &$count, $level) use ($meta, $config, $self, $em, $options, $updateQb, &$doRecover) {
+            $rootEntity = $em->getReference($meta->getName(), $root['node_id']);
+            $lft = $count++;
```
[the remark this answers is not in the document]
Comment (pmishev) [reply src/Tree/Entity/Repository/NestedTreeRepository.php:939]: It is the `left` value of the node being processed

File: src/Tree/Mapping/Driver/Annotation.php [hunk 9]
```diff
@@ -180,7 +180,7 @@ public function readExtendedMetadata($meta, array &$config)
                 $config['root'] = $field;
             }
             // level
-            if ($this->reader->getPropertyAnnotation($property, self::LEVEL)) {
+            if ($levelAnnotation = $this->reader->getPropertyAnnotation($property, self::LEVEL)) {
```
[the remark this answers is not in the document]
Comment (pmishev) [reply src/Tree/Mapping/Driver/Annotation.php:183]: Do you mean to separate the assignment with the condition on 2 separate lines?
Generally I've tried to keep to the existing code style and you can see the following `if` on line 195 is done the same way.

File: src/Tree/Entity/Repository/NestedTreeRepository.php [hunk 10]
```diff
@@ -825,6 +894,10 @@ public function verify()
         if (isset($config['root'])) {
             $trees = $this->getRootNodes();
             foreach ($trees as $tree) {
+                // if a root node is specified, verify only it
+                if (is_object($options['treeRootNode']) && $options['treeRootNode'] !== $tree) {
```
Comment (phansys) [inline src/Tree/Entity/Repository/NestedTreeRepository.php:898]: If `$options['treeRootNode']` can only hold an object or `null`, I think `null !== $options['treeRootNode']` is cheaper.

File: src/Tree/Mapping/Driver/Annotation.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (phansys) [reply src/Tree/Mapping/Driver/Annotation.php:183]: Yes, but I haven't a strong argument here. Generally, the assignment inside a condition leads more changes in case of refactoring. Additionally, it makes harder to annotate the variables (when it is strictly required).

File: src/Tree/Entity/Repository/NestedTreeRepository.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (phansys) [reply src/Tree/Entity/Repository/NestedTreeRepository.php:948]: I think we can avoid this kind of abbreviations that don't save too much.

File: src/Tree/Entity/Repository/NestedTreeRepository.php [hunk 11]
```diff
@@ -835,25 +908,124 @@ public function verify()
...
+        $updateQb = $em->createQueryBuilder()
+                       ->update($meta->getName(), 'node')
+                       ->set('node.'.$config['left'], ':left')
+                       ->set('node.'.$config['right'], ':right')
+                       ->where('node.id = :id');
+        if (isset($config['level'])) {
+            $updateQb->set('node.'.$config['level'], ':level');
+        }
+
+        $doRecover = function ($root, &$count, $level) use ($meta, $em, $options, $updateQb, &$doRecover) {
```
Comment (franmomu) [inline src/Tree/Entity/Repository/NestedTreeRepository.php:946]:
```suggestion
        $doRecover = function ($root, &$count, $level) use ($meta, $em, $options, $updateQb, &$doRecover): void {
```
We could also add parameter type declarations

File: src/Tree/Mapping/Driver/Annotation.php [hunk 12]
```diff
@@ -180,7 +180,8 @@ public function readExtendedMetadata($meta, array &$config)
                 $config['root'] = $field;
             }
             // level
-            if ($this->reader->getPropertyAnnotation($property, self::LEVEL)) {
+            $levelAnnotation = $this->reader->getPropertyAnnotation($property, self::LEVEL);
+            if ($levelAnnotation) {
```
Comment (phansys) [inline src/Tree/Mapping/Driver/Annotation.php:184]: Could we use a stricter condition?

File: tests/Gedmo/Tree/NestedTreeRootRepositoryTest.php [hunk 13]
```diff
@@ -238,8 +240,36 @@ public function testShouldRemoveRootNodeFromTree(): void
...
+        $repo = $this->em->getRepository(self::CATEGORY);
+        $carrots = $repo->findOneBy(['title' => 'Carrots']);
+
+        $this->expectException(InvalidArgumentException::class);
+        $repo->getPathAsString($carrots, [
+            'stringMethod' => $stringMethod,
+        ]);
+    }
+
+    public function invalidStringMethods(): array
```
Comment (phansys) [inline tests/Gedmo/Tree/NestedTreeRootRepositoryTest.php:260]:
```suggestion
    public function invalidStringMethods(): iterable
```
You can also yield these values instead of put an array in memory.

**Review by phansys — CHANGES_REQUESTED** [review]

File: tests/Gedmo/Tree/NestedTreeRootRepositoryTest.php [hunk 14]
```diff
@@ -257,9 +257,9 @@ public function testGetPathAsStringWithInvalidStringMethod($stringMethod): void
         ]);
     }
 
-    public function invalidStringMethods(): array
+    public function invalidStringMethods(): iterable
     {
-        return [
+        yield from [
```
Comment (phansys) [inline tests/Gedmo/Tree/NestedTreeRootRepositoryTest.php:262]: If you use `yield from`, the array in the `from` part is already in memory. The real advantage of `yield` can be leveraged by yielding every single item. By instance:
```php
yield [null];
yield [123];
yield ['nonExistingMethod'];
yield [''];
```

File: tests/Gedmo/Tree/NestedTreeRootRepositoryTest.php [hunk 15]
```diff
@@ -157,7 +159,7 @@ public function testShouldSupportChildrenHierarchyAsHtml(): void
         $childOpen = '';
         $childClose = '';
         $nodeDecorator = static function ($node) {
-            return str_repeat('-', $node['level']).$node['title']."\n";
+            return str_repeat('-', $node['level'] - 1).$node['title']."\n";
```
Comment (franmomu) [inline tests/Gedmo/Tree/NestedTreeRootRepositoryTest.php:162]: is this change needed because the level of `RootCategory` has been changed from 0 to 1? if that's the case I think we should change the expected output instead of this or do not change `RootCategory` level and create another fixture to test this new feature.

File: tests/Gedmo/Tree/NestedTreeRootRepositoryTest.php [hunk 16]
```diff
@@ -174,7 +176,7 @@ public function testShouldSupportChildrenHierarchyAsHtml(): void
         // check support of the closures in rootClose
         $rootClose = static function () {return '</ul><!--rootCloseClosure-->'; };
         $childOpen = static function (&$node) {
-            return '<li class="depth'.$node['level'].'">';
+            return '<li class="depth'.($node['level'] - 1).'">';
```
Comment (franmomu) [inline tests/Gedmo/Tree/NestedTreeRootRepositoryTest.php:179]: Here the same thing than https://github.com/doctrine-extensions/DoctrineExtensions/pull/2447/files#r922696534

File: tests/Gedmo/Tree/NestedTreeRootRepositoryTest.php (hunk 15 above)
Comment (pmishev) [reply tests/Gedmo/Tree/NestedTreeRootRepositoryTest.php:162]: Yes, that's the case. Changing the root level to 1 would change the test output from:
```
-Fruits
-Vegitables
--Carrots
--Potatoes
```
to:
```
--Fruits
--Vegitables
---Carrots
---Potatoes
```
I decided to change the node decorator to keep the old output as it looks more sensible. Other than that, I don't think it really matters if I changed the output instead. But, if you prefer that, I can change the output.
As for creating another fixture, I agree that would be ideal, but I've had to change the expected level in multiple tests that use the RootCategory fixture and if I had to to another fixture, this would result in a lot of duplicated tests. It just seemed a bit too much. If that's the way you prefer to go though, I'll need some time to work on it.

Comment (pmishev) [reply tests/Gedmo/Tree/NestedTreeRootRepositoryTest.php:162]: Just to give you some context on this feature and the point of it - it was necessary because we were incrementally migrating from another framework, working with a legacy DB that we couldn't change at the time.

Comment (franmomu) [reply tests/Gedmo/Tree/NestedTreeRootRepositoryTest.php:162]: It's not a blocker, I was just a bit concerned about someone reading the code and do not understand that that `-1` inside the method is because `RootCategory` has level 1, I guess is like in the other changes it would be:

```diff
- static::assertSame(1, $node->getLevel());
+ static::assertSame(1, $node->getLevel() - 1);
```
Anyway, it's fine by me when it's rebased.

**Review by franmomu — APPROVED** [review]

File: src/Tree/Entity/Repository/NestedTreeRepository.php [hunk 17]
```diff
@@ -172,24 +187,69 @@ public function getPathQueryBuilder($node)
      * Get the Tree path query by given $node
      *
      * @param object $node
+     * @phpstan-param array{includeNode?: bool} $options
+     *
+     * $options['includeNode'] = (bool) Whether to include the node itself (default: true)
```
Comment (phansys) [inline src/Tree/Entity/Repository/NestedTreeRepository.php:192]: If you need to add a description for the method arguments, please use the conventional PHPDoc `@param` annotation:
```suggestion
     * @param array $options
     *    bool "includeNode": Whether to include the node itself (default: true)
     *
     * @phpstan-param array{includeNode?: bool} $options
```

File: src/Tree/Entity/Repository/NestedTreeRepository.php [hunk 18]
```diff
@@ -172,24 +187,69 @@ public function getPathQueryBuilder($node)
...
      *
-     * @return \Doctrine\ORM\Query
+     * @return Query
      */
-    public function getPathQuery($node)
+    public function getPathQuery($node, array $options = [])
```
Comment (phansys) [inline src/Tree/Entity/Repository/NestedTreeRepository.php:196]: This is a BC break.

File: src/Tree/Entity/Repository/NestedTreeRepository.php [hunk 19]
```diff
@@ -61,7 +63,7 @@ public function __call($method, $args)
     {
         if ('persistAs' === substr($method, 0, 9)) {
             if (!isset($args[0])) {
-                throw new \Gedmo\Exception\InvalidArgumentException('Node to persist must be available as first argument');
+                throw new InvalidArgumentException('Node to persist must be available as first argument');
```
Comment (phansys) [inline src/Tree/Entity/Repository/NestedTreeRepository.php:66]:
```suggestion
                throw new InvalidArgumentException('Node to persist must be available as first argument.');
```

File: src/Tree/Entity/Repository/NestedTreeRepository.php [hunk 20]
```diff
@@ -70,7 +72,7 @@ public function __call($method, $args)
             $position = substr($method, 9);
             if ('Of' === substr($method, -2)) {
                 if (!isset($args[1])) {
-                    throw new \Gedmo\Exception\InvalidArgumentException('If "Of" is specified you must provide parent or sibling as the second argument');
+                    throw new InvalidArgumentException('If "Of" is specified you must provide parent or sibling as the second argument');
```
Comment (phansys) [inline src/Tree/Entity/Repository/NestedTreeRepository.php:75]:
```suggestion
                    throw new InvalidArgumentException('If "Of" is specified you must provide parent or sibling as the second argument.');
```

File: src/Tree/Entity/Repository/NestedTreeRepository.php [hunk 21]
```diff
@@ -172,24 +187,69 @@ public function getPathQueryBuilder($node)
...
+    {
+        $defaultOptions = [
+            'includeNode' => true,
+            'separator' => ' > ',
+            'stringMethod' => '__toString',
+        ];
+        $options += $defaultOptions;
+
+        if (!is_string($options['stringMethod'])) {
+            throw new InvalidArgumentException(sprintf('"stringMethod" parameter passed to %s must be a valid string', __METHOD__));
```
Comment (phansys) [inline src/Tree/Entity/Repository/NestedTreeRepository.php:241]:
```suggestion
            throw new InvalidArgumentException(sprintf('"stringMethod" option passed in argument 2 to %s must be a valid string.', __METHOD__));
```

File: src/Tree/Entity/Repository/NestedTreeRepository.php [hunk 22]
```diff
@@ -172,24 +187,69 @@ public function getPathQueryBuilder($node)
...
+        }
+        if (!method_exists($node, $options['stringMethod'])) {
+            throw new InvalidArgumentException(sprintf('%s must implement method "%s"', get_class($node), $options['stringMethod']));
```
Comment (phansys) [inline src/Tree/Entity/Repository/NestedTreeRepository.php:244]:
```suggestion
            throw new InvalidArgumentException(sprintf('%s must implement method "%s".', get_class($node), $options['stringMethod']));
```

**Review by phansys — CHANGES_REQUESTED** [review]

File: src/Tree/Entity/Repository/NestedTreeRepository.php (hunk 17 above)
Comment (pmishev) [reply src/Tree/Entity/Repository/NestedTreeRepository.php:192]: Ok, I wasn't aware there was a convention in documenting these. For example here they are described a bit differently: https://github.com/doctrine-extensions/DoctrineExtensions/blob/c51e05ebbeccdabe9d0f09cc264c1657880a2bfa/src/Tree/RepositoryUtilsInterface.php#L47
Your proposed format looks nice, so I can change my comments like that. Do you mind that there will still be a discrepancy with the above example?

File: src/Tree/Entity/Repository/NestedTreeRepository.php (hunk 18 above)
Comment (pmishev) [reply src/Tree/Entity/Repository/NestedTreeRepository.php:196]: I'm sorry, how is this a BC break?

File: src/Tree/Entity/Repository/NestedTreeRepository.php (hunk 19 above)
Comment (pmishev) [reply src/Tree/Entity/Repository/NestedTreeRepository.php:66]: Doesn't this qualify as a BC change, in the unlikely scenario someone is checking for the exception message text?

File: src/Tree/Entity/Repository/NestedTreeRepository.php (hunk 18 above)
Comment (phansys) [reply src/Tree/Entity/Repository/NestedTreeRepository.php:196]: This is adding a new parameter to a method that (currently) can be extended by users.
See:
* https://3v4l.org/ldmIb;
* https://en.wikipedia.org/wiki/Liskov_substitution_principle.

File: src/Tree/Entity/Repository/NestedTreeRepository.php (hunk 19 above)
Comment (phansys) [reply src/Tree/Entity/Repository/NestedTreeRepository.php:66]: I honestly hope not. But if you think it may be the case, please dismiss my suggestion.

File: src/Tree/Entity/Repository/NestedTreeRepository.php (hunk 17 above)
Comment (phansys) [reply src/Tree/Entity/Repository/NestedTreeRepository.php:192]: I think the only implicit convention is to leverage the PHPDoc annotations when possible. We have a hard normalizing the existing codebase. By now, we haven't a rule for the indentation other than consistency, so feel free to come up the way you think is a better fit.

Comment (pmishev) [reply src/Tree/Entity/Repository/NestedTreeRepository.php:192]: in this case, I'd like to suggest:
```
     * @param array $options
     *     includeNode => (bool) Whether to include the node itself (default: true),
```
I think it's a bit more readable having the param name first, not the type, if you have many params and you're trying to scan through them. WDYT?

File: src/Tree/Entity/Repository/NestedTreeRepository.php (hunk 18 above)
Comment (pmishev) [reply src/Tree/Entity/Repository/NestedTreeRepository.php:196]: Well yes, you have a point, but if we're going to consider this a breaking change, I have also added new methods, such as `getPathAsString()` and `recoverFast()`, so they are then breaking changes as well. What do you suggest to do about that?

File: src/Tree/Entity/Repository/NestedTreeRepository.php (hunk 17 above)
Comment (phansys) [reply src/Tree/Entity/Repository/NestedTreeRepository.php:192]: I have no personal preference here. The only thing I can use as base for the syntax I've suggested is the structure used in the PHPDoc [`@param`](https://docs.phpdoc.org/3.0/guide/references/phpdoc/tags/param.html) tag:
```
@param [<Type>] [name] [<description>]
```

File: src/Tree/Entity/Repository/NestedTreeRepository.php (hunk 18 above)
Comment (phansys) [reply src/Tree/Entity/Repository/NestedTreeRepository.php:196]: AFAIK, adding new methods in a class that is not part of the current API is not a BC break.

Comment (franmomu) [reply src/Tree/Entity/Repository/NestedTreeRepository.php:196]: you can do something similar to: 

```php
$options = func_get_args()[1] ?? [];

if (!\is_array($options)) {
    throw new TypeError('Argument 2 MUST be an array');
}
```
And we'll add the parameter in the next major version.

File: src/Tree/Entity/Repository/NestedTreeRepository.php (hunk 17 above)
Comment (pmishev) [reply src/Tree/Entity/Repository/NestedTreeRepository.php]: I updated the phpdocs according to this https://stackoverflow.com/a/51032174/2737626

File: src/Tree/Entity/Repository/NestedTreeRepository.php (hunk 18 above)
Comment (pmishev) [reply src/Tree/Entity/Repository/NestedTreeRepository.php:200]: Are you suggesting that I create a separate PR where the only changes are adding the above snippet in the following methods?
```
    public function getPathQueryBuilder($node)
    public function getPathQuery($node)
    public function getPath($node)
    public function verify()
    public function recover()
```
...and then this PR will just wait for a next major version?

TBH, I'm not sure what help will this snippet do. Someone might have extended the method and added a second array argument, in which case they won't get a TypeError and in the next major version, their code will still be broken.

Comment (pmishev) [reply src/Tree/Entity/Repository/NestedTreeRepository.php]: Apologies, I've been away for a while. I've made the requested changes now.

**Comment by pmishev** [comment]
Just checking in to see if you need anything more fixed in this PR?

**Comment by franmomu** [comment]
> Just checking in to see if you need anything more fixed in this PR?

Just a rebase to fix the conflict in the `CHANGLOG:md` file from my side, let see if @phansys can give another review.

**Review by phansys — COMMENTED** [review]
I'm not convinced about the `@param` annotations like this:
```
 * @param array  $options = [
 *                        'includeNode' => true, // Whether to include the node itself
 *                        ]
 *
```
but I don't want to block the additions introduced in this PR.

**Comment by pmishev** [comment]
Did the rebase. 
As far as the `@param` are concerned, we had a discussion about those above. I did some research then and didn't find any consensus on what's the best way to document those, so I picked one that seemed the least bad. I don't really care much about those, so if you prefer another one, I can change them.

**Comment by phansys** [comment]
>I don't really care much about those, so if you prefer another one, I can change them.

For lack of a better example, I think [this one](https://github.com/aws/aws-sdk-php/blob/5791e496e3fc0847f40c86515ed8659b0ecd5991/src/S3/S3Client.php#L297-L345) a bit tidier.

**Comment by pmishev** [comment]
So, instead of this:
```
     * @param object $node
     * @param array  $options = [
     *                        'includeNode' => true, // Whether to include the node itself
     *                        ]
     *
     * @phpstan-param array{includeNode?: bool} $options
```
it would be this:
```
     * @param object $node
     * @phpstan-param array{includeNode?: bool} $options
     *
     * options:
     * - includeNode: (bool) Whether to include the node itself. Defaults to true.
```

Just to point out the cons I see in this format:
- The default values are not immediately obvious.
- The `$option` will no longer have a `@param` annotation

Please confirm and I'll change them.

**Comment by phansys** [comment]
Thanks @pmishev.
As I mentioned, I don't want to block this PR.
Regarding your points, I guess we can keep the `@param` tag if it adds some value, like `@param array<string, mixed> $options`.
About the default value, I find the example explicit enough.

**Review by phansys — APPROVED** [review]

**Comment by franmomu** [comment]
thanks for keeping working on this @pmishev! and sorry for taking that long
